api/v1alpha4/machinedeployment_types.go

```diff
@@ -21,6 +21,12 @@ import (
 	"k8s.io/apimachinery/pkg/util/intstr"
 )
 
+const (
+	// MachineDeploymentTopologyFinalizer is the finalizer used by the topology MachineDeployment controller to
+	// clean up referenced template resources if necessary when a MachineDeployment is being deleted.
```

Comment:
Do these templates have owner refs? If so they would just go orphan and garbage collected so no need to explicitly take care of them?

Reply:
They have ownerReferences but we currently only set the cluster. The idea of that is that in the regular non-topology managed case templates can be reused between MachineDeployments and are automatically deleted together with the cluster. So with only this mechanism we would collect templates until the cluster is deleted.

If we want to solve the topology-owned case with ownerReference we would have to:
* make sure that in this case we don't set the cluster as ownerReference
* make sure that we always have the right ownerReferences in the template resource, i.e. we have to update the ownerReferences when the templates in the MD or MS change
    * I think for that we basically also have to iterate over everything because we have to ensure it also works when we miss an individual update event, which can always happen.

So I think it wouldn't be simpler and we would have to change the existing controllers to behave differently for topology-owned MD/MS to achieve that.

Reply:
Changing the owner ref chain is probably a breaking change and it can also have subtle implications, eg. for clusterctl move or for infrastructure machines which now might assume on a certain deletion order.

I would suggest to open an issue to track this idea, while this PR should provide a solution without changing the current deletion process.

Reply:
Also, the assumption that a template is linked to a single MD/MS holds only for managed topologies

Reply:
>make sure that we always have the right ownerReferences in the template resource, i.e. we have to update the ownerReferences when the templates in the MD or MS change
I think for that we basically also have to iterate over everything because we have to ensure it also works when we miss an individual update event, which can always happen.

Good point. I'd expect though whatever controller is doing the rotation to just pick the current template and remove the MachineDeployment owner.
Also if we set only MachineSet owner refs the templates would just be handled automatically as MachineDeployments manages/delete MachineSets, there'd be no need to remove any owner from a higher level. I think that'd be the simplest solution without accounting for Fabrizio concerns about clusterctl.

>Also, the assumption that a template is linked to a single MD/MS holds only for managed topologies
Agree, true for any approach we go with.

sgtm continuing with this approach, but definitely worth exploring if we can save all this code by setting an owner ref.

Reply:
I'm not sure the simplified approach would work.

It relies on that you always have a MachineSet for a MachineDeployment. For example, (afaik) if you scale a MD down to `0` you will only have at most a `RevisionHistoryLimit` number of  MachineSets. If `RevisionHistoryLimit` is `0` (default is `1`), you will have a MD with referenced templates without a MachineSet. So this means the templates would be deleted. This also means when you scale up the MachineDeployment again you have to make sure to create the templates again.